### PR TITLE
Enhanced the look and usability of Keymap editor.

### PR DESCRIPTION
**Pull request by TeaCrab**
What's changed:
- Changing settings in keymap used to occasionally fold(un-expand) the list of entries.  Now it stays in place, making keymap editing less annoying to deal with.
- Changed the search/filter function to work with multiple words separated by whitespace.  Added some convenient short handles for common keys such as "SS" being recognized as the shift key.
- Keymap entries are now displayed a bit nicer to the eyes.

### CODE REVIEW

File: release/scripts/modules/rna_keymap_ui.py [hunk 1]
```diff
@@ -122,11 +122,8 @@ def draw_kmi(display_keymaps, kc, km, kmi, layout, level):
 
     col = _indented_layout(layout, level)
 
-    if kmi.show_expanded:
-        col = col.column(align=True)
-        box = col.box()
-    else:
-        box = col.column()
+    col = col.column(align=True)
+    box = col.box()
```
Comment (TeaCrab) [inline release/scripts/modules/rna_keymap_ui.py]: I think, this part solved the issue of "Unexpand Upon Change" behavior I have regularly encountered in the past.  Don't know if anyone else had the same issue.

The problem might have been that, depending on this `kmi.show_expanded` value, different types of UI elements are generated.  And somehow also resets this `kmi.show_expanded` value.  This is only one of the assumptions for why the "Unexpand Upon Change" behavior.

By eliminating the logic to create 2 streams of possibility, the problem seems to be fixed.

File: release/scripts/modules/rna_keymap_ui.py [hunk 2]
```diff
@@ -165,9 +162,9 @@ def draw_kmi(display_keymaps, kc, km, kmi, layout, level):
 
     # Expanded, additional event settings
     if kmi.show_expanded:
```
Comment (TeaCrab) [inline release/scripts/modules/rna_keymap_ui.py]: This is what I think, the old UI logic essentially starts to fall apart and created the "Unexpand Upon Change" behavior.  The code changes after this are mostly visual changes made to make keymap editor look nicer.

File: release/scripts/modules/rna_keymap_ui.py [hunk 3]
```diff
@@ -122,11 +122,8 @@ def draw_kmi(display_keymaps, kc, km, kmi, layout, level):
 
     col = _indented_layout(layout, level)
 
-    if kmi.show_expanded:
-        col = col.column(align=True)
```
Comment (TeaCrab) [inline release/scripts/modules/rna_keymap_ui.py]: This is the part that potentially created the "Unexpand Upon Change" behavior.

File: release/scripts/modules/rna_keymap_ui.py [hunk 4]
```diff
@@ -212,13 +209,12 @@ def draw_kmi(display_keymaps, kc, km, kmi, layout, level):
...
-
 def draw_filtered(display_keymaps, filter_type, filter_text, layout):
-
     if filter_type == 'NAME':
         def filter_func(kmi):
-            return (filter_text in kmi.idname.lower() or
-                    filter_text in kmi.name.lower())
+            compiled_filter = filter_text.split(' ')
+            return all(x in kmi.idname.lower() for x in compiled_filter ) \
+            or all(x in kmi.name.lower() for x in compiled_filter)
```
Comment (TeaCrab) [inline release/scripts/modules/rna_keymap_ui.py]: Multi-word filter.

File: release/scripts/modules/rna_keymap_ui.py [hunk 5]
```diff
@@ -228,13 +224,25 @@ def filter_func(kmi):
...
+                "RMB" : 'RIGHTMOUSE'     ,
+                "LMB" : 'LEFTMOUSE'      ,
+                "MMB" : 'MIDDLEMOUSE'    ,
+                "DA"  : 'LEFT_ARROW'     ,
+                "DD"  : 'RIGHT_ARROW'    ,
+                "DW"  : 'UP_ARROW'       ,
+                "DS"  : 'DOWN_ARROW'     ,
+                "PU"  : 'PAGE_UP'        ,
+                "PD"  : 'PAGE_DOWN'      ,
+
```
Comment (TeaCrab) [inline release/scripts/modules/rna_keymap_ui.py]: Just silly me adding short handles for the filter bar.

**Comment by dfelinto** [comment]
Hi, please use developer.blender.org for patches and contributions to Blender. This repository is my personal repository for backup :)

**Comment by TeaCrab** [comment]
Yea realized that.  Will use  developer.blender.org  instead.

Brent Liu
Concept & 3D Artist
Link for portfolio <https://www.artstation.com/artist/remotecrab131>

**Pull request closed without merging**